Start Requiring Zulip Server 5

PIG208 opened the pull request:
Blocked by #267, but these are migrations that can happen in parallel. I imagine that there won't be much overhead rebasing this kind of changes.

Fixes: #268

PIG208 commented:
The PR itself is ready, but we won't be ready to merge it until we refuse to connect to older servers. Going through some reviews will still be helpful, though.

chrisbobbe commented on lib/api/model/model.dart:
```diff
@@ -205,7 +205,7 @@ class User {
   // bool isAdmin; // obsoleted by [role]; ignore
   // bool isGuest; // obsoleted by [role]; ignore
-  bool? isBillingAdmin; // TODO(server-5)
+  bool isBillingAdmin;
```
```
api: Mark isBillingAdmin as requried, relying on server 5+, FL 73+

While `realm_user` update events also have this field added at the same
feature level, it remains optional.

See "Feature level 73" from Zulip API changelog:

   https://zulip.com/api/changelog

Signed-off-by: Zixuan James Li <zixuan@zulip.com>
```

I would just leave out the sentence about how `isBillingAdmin` remains optional in `RealmUserUpdateEvent`. That's the normal, expected thing to happen; it matches all the other `person` fields, in that they might be null if they didn't change in the event.

Also nit in summary line: 'required"

chrisbobbe commented on lib/api/model/model.dart:
```diff
@@ -233,14 +233,6 @@ class User {
   }
 
-  static bool _readIsSystemBot(Map<dynamic, dynamic> json, String key) {
-    // This field is absent in `realm_users` and `realm_non_active_users`,
-    // which contain no system bots; it's present in `cross_realm_bots`.
```
```
api: Drop _readIsSystemBot, relying on server 5+, FL 83+
```

This comment still applies, right? The change made in FL 83 is just that `is_cross_realm_bot` stops being a name we have to care about. The comment can go above the `defaultValue: false`, and it seems like we should keep the test that confirms that `false` is chosen if `is_system_bot` is absent.

chrisbobbe commented on lib/api/model/model.dart:
```diff
@@ -824,16 +824,12 @@ enum MessageEditState {
-        }
+      if (topic != null || prevTopic != null) {
+        // Both are present if the topic was edited
+        assert(topic != null && prevTopic != null);
```
The condition in this `assert` might happen because of unexpected data from the server. To validate data from the server, we'll want code that runs in production, and reserve `assert`s to check our own invariants that we maintain within the bounds of the app. Could replace the assert and comment with:

```dart
        // Crunch-shell validation: both are present if the topic was edited
        topic as String;
        prevTopic as String;
```

chrisbobbe reviewed (commented):
Thanks for going ahead with this! This will be good to have ready to merge once #267 is done (that's on my plate right now). Small comments below.

PIG208 commented:
Thanks! Updated the PR.

chrisbobbe commented:
Thanks! LGTM except bump on this part of https://github.com/zulip/zulip-flutter/pull/904#discussion_r1732067432 :

> it seems like we should keep the test that confirms that false is chosen if `is_system_bot` is absent.

PIG208 commented:
Got that test back. Thanks for the review!

chrisbobbe commented:
Great! Marking for Greg's review once he's back from vacation.

gnprice commented:
BTW there's generally no need to push a revision when it's a pure rebase, with no merge conflicts or other changes. The update causes notifications, and doesn't really save effort later — there'll typically be another rebase anyway just before merge, and `git rebase` is a very easy step to do at merge time when there's no conflict to resolve.

(As it happens I'm just about to take a look at this PR, but it wasn't because of the notification from the revision just now; it was from scanning the list of [PRs marked for my review](https://github.com/zulip/zulip-flutter/pulls?q=is%3Apr+is%3Aopen+label%3A%22integration+review%22).)

gnprice commented on lib/api/route/messages.dart:
```diff
@@ -1,64 +1,16 @@
-///
-/// Gives null if the server reports that the message doesn't exist.
-// TODO(server-5) Simplify this away; just use getMessage.
-Future<Message?> getMessageCompat(ApiConnection connection, {
```
nit in commit message:

    See also: 631f4d68c58a84dff66a6a9a80322584f0d60152

Instead say:

    See also commit 631f4d68c58a84dff66a6a9a80322584f0d60152.

That way it's clear this hex blob refers to a commit (implicitly one in this same repo).

gnprice reviewed (commented):
Thanks @PIG208 for taking care of these, and thanks @chrisbobbe for the previous reviews!

Generally this looks great. Small comments below.

PIG208 commented on lib/api/model/model.dart:
```diff
@@ -205,7 +205,7 @@ class User {
   // bool isAdmin; // obsoleted by [role]; ignore
   // bool isGuest; // obsoleted by [role]; ignore
-  bool? isBillingAdmin; // TODO(server-5)
+  bool isBillingAdmin;
```
Yeah, I think the context info is pretty helpful. While the commit summaries get longer, it is justified for this kind of migration changes.

PIG208 commented:
Thanks for the review! Added some TODOs commits for `GetServerSettingsResult` changes, making them easy to drop, and addressed the issues from the review. This revision should be fairly readable with range-diff.

gnprice commented on lib/api/route/realm.g.dart:
```diff
@@ -20,7 +20,7 @@ GetServerSettingsResult _$GetServerSettingsResultFromJson(
       zulipFeatureLevel: (json['zulip_feature_level'] as num).toInt(),
       zulipVersion: json['zulip_version'] as String,
-      zulipMergeBase: json['zulip_merge_base'] as String?,
+      zulipMergeBase: json['zulip_merge_base'] as String,
```
This doesn't seem to have a corresponding change in realm.dart.

Presumably this came about as you made some edits and missed rerunning `build_runner` for one of them, or something like that. I am puzzled that our CI didn't complain, though — `tools/check build_runner` should have failed because this doesn't match what a freshly-generated file should say. So that would be worth trying to debug.

PIG208 replied:
For future reference, this was fixed in #947.

gnprice commented on lib/api/route/realm.dart:
```diff
@@ -34,7 +34,8 @@ class GetServerSettingsResult {
   final int zulipFeatureLevel;
   final String zulipVersion;
-  final String? zulipMergeBase; // TODO(server-5)
+  // TODO: Modernize this once we get to #267
+  final String? zulipMergeBase;
```
This should keep a `TODO(server-5)` comment until we do actually modernize it. That way it naturally shows up in any future round of grepping for opportunities to modernize our API usage.

gnprice commented on lib/api/model/web_auth.dart:
```diff
@@ -7,7 +7,7 @@ import 'package:flutter/foundation.dart';
   final Uri realm;
   final String email;
-  final int? userId; // TODO(server-5) new in FL 108
+  final int userId;
```
bump / nit: use `api:` for prefix, like for other API changes

(in this revision it's:

    login: Mark WebAuthPayload.userId as required, relying on server 5+, FL 108+.

)

gnprice commented on lib/api/model/events.dart:
```diff
@@ -704,13 +704,13 @@ class UpdateMessageEvent extends Event {
 
   final List<MessageFlag> flags;
-  final int? editTimestamp; // TODO(server-5)
+  final int editTimestamp;
```
nit: this doesn't seem accurate:

    Fields including `userId` and `editTimestamp` are optional and act
    as alternatives to `renderingOnly` for older server versions.

since those fields are no longer optional.

I think the commit message can skip explaining the fallback that previously existed for renderingOnly. It's directly relevant only when looking at the past; and the details are in this commit's diff.

PIG208 replied:
Yeah, "were" would be more accurate here. Leaving this out makes sense.

gnprice commented on lib/api/model/initial_snapshot.dart:
```diff
@@ -275,15 +269,9 @@ class UnreadMessagesSnapshot {
 /// An item in [UnreadMessagesSnapshot.dms].
 @JsonSerializable(fieldRename: FieldRename.snake)
 class UnreadDmSnapshot {
-  @JsonKey(readValue: _readOtherUserId)
   final int otherUserId;
```
nit:

    api: Remove sender_id fallback for UnreadMessagesSnapshot.otherUserId, relying on server 5+, FL 119+

that's not the class this field is on :slightly_smiling_face:

PIG208 replied:
Good find! I should learn to not trust the diffs heading and be more careful from now on :sweat: 

```diff
diff --git a/lib/api/model/initial_snapshot.dart b/lib/api/model/initial_snapshot.dart
index 565d89fe..6be29e78 100644
--- a/lib/api/model/initial_snapshot.dart
+++ b/lib/api/model/initial_snapshot.dart
@@ -269,15 +269,9 @@ class UnreadMessagesSnapshot {
 /// An item in [UnreadMessagesSnapshot.dms].
 @JsonSerializable(fieldRename: FieldRename.snake)
 class UnreadDmSnapshot {
-  @JsonKey(readValue: _readOtherUserId)
   final int otherUserId;
   final List<int> unreadMessageIds;
 
-  // TODO(server-5): Simplify away.
-  static dynamic _readOtherUserId(Map<dynamic, dynamic> json, String key) {
-    return json[key] ?? json['sender_id'];
-  }
-
```

gnprice replied:
Ah, yeah, the diff-hunk headers can be misleading like that sometimes :slightly_smiling_face:

gnprice reviewed (commented):
Thanks for the revisions! Just a few comments below.

gnprice commented:
Also marked as draft to reflect that we won't be ready to merge this until we have #267.

Marking as draft helps avoid me forgetting that and going ahead and merging after a next round of review, if the thread otherwise looks ready :slightly_smiling_face:

gnprice commented:
Removed the label to get this out of my queue since it's on hold (as described above). We'll bring it back when we pick up the underlying issue again.

gnprice commented:
We have #267 now, so this is something we could potentially merge when its code is ready. I've just rebased it atop current main.

chrisbobbe commented on lib/api/route/realm.dart:
```diff
@@ -36,7 +36,8 @@ class GetServerSettingsResult {
   final int zulipFeatureLevel;
   final String zulipVersion;
-  final String? zulipMergeBase; // TODO(server-5)
+  // TODO(server-5): Modernize this once we get to #267.
+  final String? zulipMergeBase;
```
```
TODO api: Add TODO to modernize GetServerSettingsResult.zulipMergeBase

Marked as TODO as this might change in the future.
See https://github.com/zulip/zulip-flutter/pull/904#discussion_r1747943940.

See "Feature level 88" from Zulip API changelog:

   https://zulip.com/api/changelog

Signed-off-by: Zixuan James Li <zixuan@zulip.com>
```

I read the linked discussion, and this change can go ahead; it doesn't interfere with what we've implemented for disallow-connecting-to-ancient-servers.

(Ditto the corresponding change in the initial snapshot, which would have been more likely to have a conflict but also doesn't.)

chrisbobbe commented on lib/api/model/events.dart:
```diff
@@ -797,13 +797,13 @@ class UpdateMessageEvent extends Event {
 
-  final int? userId; // TODO(server-5)
-  final bool? renderingOnly; // TODO(server-5)
+  final int userId;
```
Discussion on this `userId` change: [#api documentation > &#96;user_id&#96; in &#96;update_message&#96; event @ 💬](https://chat.zulip.org/#narrow/channel/412-api-documentation/topic/.60user_id.60.20in.20.60update_message.60.20event/near/2215174)

chrisbobbe commented on lib/api/route/realm.dart:
```diff
@@ -51,7 +51,7 @@ class GetServerSettingsResult {
   final String realmIcon;
   final String realmDescription;
-  final bool? realmWebPublicAccessEnabled; // TODO(server-5)
+  final bool realmWebPublicAccessEnabled;
```
```
TODO api: Mark GetServerSettingsResult.realmWebPublicAccessEnabled as required, relying on server 5+, FL 116+

See https://github.com/zulip/zulip-flutter/pull/904#discussion_r1747943940.

See "Feature level 116" from Zulip API changelog:

   https://zulip.com/api/changelog

Signed-off-by: Zixuan James Li <zixuan@zulip.com>
```

as in a previous comment of mine: this is fine; no interaction with disallow-ancient-servers

chrisbobbe reviewed (commented):
Thanks @PIG208!

Here's a review, since this is marked for maintainer review—but probably Greg or I will end up making new revisions?

There are three `TODO(server-5)` comments left at the tip of the branch; I've commented on one in my review, and the other two are also small but snuck in after this PR was first written almost a year ago.

PIG208 commented:
Pushed an update addressing the new TODOs and the comments above.

gnprice commented on lib/api/model/web_auth.dart:
```diff
@@ -25,21 +25,16 @@ class WebAuthPayload {
-        userId = int.tryParse(userIdStr, radix: 10);
-        if (userId == null) throw const FormatException();
-      }
+      final int? userId = int.tryParse(userIdStr, radix: 10);
```
… I guess the existing code just above for Uri.tryParse does the same thing, so this matches. Whatever, then; we can clean up them both if we're touching this code in the future.
